File: LeetCode/Explore/LinkedList/DetectCycle.cpp
#include "../main.h"

class Solution {
  public:
    ListNode *detectCycle(ListNode *head) {
        unordered_set<ListNode *> st;
        auto p = head;
        while (p) {
            if (st.insert(p).second == false)
                return p;
            p = p->next;
        }
        return nullptr;
    }
};

File: LeetCode/LCCI/detectCycle.go
package lcci

func detectCycle(head *ListNode) *ListNode {
	p := head
	set := map[*ListNode]bool{}
	for p != nil {
		if set[p] == true {
			return p
		}
		set[p] = true
		p = p.Next
	}
	return nil
}
